File: NinePSharp/messages/Rwrite.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Rwrite tag[2] count[4]
public readonly struct Rwrite : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rwrite;
    public ushort Tag { get; }

    public uint Count { get; }

    public Rwrite(ushort tag, uint count)
    {
        Tag = tag;
        Count = count;
        Size = NinePConstants.HeaderSize + 4;
    }

    public Rwrite(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;
        Count = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Count);
    }
}

File: NinePSharp.Messages.FSharp/messages/Rwrite.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rwrite =
    { Size: uint32
      Tag: uint16
      Count: uint32 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rwrite
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rwrite span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Count)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let count = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        { Size = size; Tag = tag; Count = count }

